### PR TITLE
DataViews: Add sync filter in patterns page

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
This PR adds a `sync status` filter in patterns page. 

This PR does a couple of more small things:
1. Adds the ability for elements of fields with type: `enumeration` to provide a helping description
2. Adds an extra check to not a show a filter if no elements are provided. Right now it would render an empty Dropdown

### Design wise
The old patterns page adds an extra header when we apply one of the `sync status` filter. For now I just added that description in the filter menu item, as a helper text. 

## Testing Instructions
1. Enable the admin views experiment and go to patterns page
3. Check that `sync status` filters works as expected
4. Observe that `sync status` field is not visible in sorting and hideable fields menus

## Screenshots or screencast 
https://github.com/WordPress/gutenberg/assets/16275880/28ede58d-ed95-48f1-9534-818f876617f5

### CODE REVIEW

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js [hunk 1]
```diff
@@ -235,12 +253,23 @@ export default function DataviewsPatterns() {
...
 			},
-		],
-		[ view.type, categoryId ]
-	);
+		];
+		if ( type === PATTERN_TYPES.theme ) {
+			_fields.push( {
+				header: __( 'Sync Status' ),
+				id: 'sync-status',
+				isVirtual: true,
```
Comment (youknowriad) [inline packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: Why should this be a virtual filter. For me it's actually a valid field that can show Sync or not.

**Review by youknowriad — COMMENTED** [review]
This is looking good other than the fact that I'm not sure there's any valid reason to make this a "virtual" field.

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js (hunk 1 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: We could definitely delay this `virtual` concept for now, and by not providing a `render` prop and adjusting the views to check for that, we could not display the field. My exploration was based on thoughts I had for other cases too, as I describe in the PR's description.

Comment (youknowriad) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: To be honest, I can't think of a case where a filter doesn't make sense as a regular field personally.

File: packages/dataviews/src/view-grid.js [hunk 2]
```diff
@@ -30,6 +30,7 @@ export default function ViewGrid( {
 	);
 	const visibleFields = fields.filter(
 		( field ) =>
+			!! field.render &&
```
Comment (youknowriad) [inline packages/dataviews/src/view-grid.js]: Why this change?

File: packages/dataviews/src/view-list.js [hunk 3]
```diff
@@ -32,6 +32,7 @@ export default function ViewList( {
 	);
 	const visibleFields = fields.filter(
 		( field ) =>
+			!! field.render &&
```
Comment (youknowriad) [inline packages/dataviews/src/view-list.js]: Why this check?

File: packages/dataviews/src/view-grid.js (hunk 2 above)
Comment (ntsekouras) [reply packages/dataviews/src/view-grid.js]: Not to render anything. In current patterns page we don't display any info for that.

>Since we don't want to display something for this field and filters are tied to fields, this PR achieves that by not providing a render function and adding the extra checks in views for this method's existence.

Comment (youknowriad) [reply packages/dataviews/src/view-grid.js]: I think a field without a render function is just a field that should render its value directly. And sure in the default view, we don't want to render the "sync status", but I don't see why this shouldn't be something users can display?

Comment (ntsekouras) [reply packages/dataviews/src/view-grid.js]: Sure, I updated the code.

File: packages/dataviews/src/view-table.js [hunk 4]
```diff
@@ -200,6 +201,11 @@ function HeaderMenu( { field, view, onChangeView } ) {
 												<DropdownMenuItemLabel>
 													{ element.label }
 												</DropdownMenuItemLabel>
+												{ !! element.description && (
```
Comment (oandregal) [inline packages/dataviews/src/view-table.js:204]: Thanks for implementing the description also for the table filters, despite this layout not being used in patterns.

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js [hunk 5]
```diff
@@ -235,12 +253,33 @@ export default function DataviewsPatterns() {
...
+		];
+		if ( type === PATTERN_TYPES.theme ) {
+			_fields.push( {
+				header: __( 'Sync Status' ),
+				id: 'sync-status',
+				render: ( { item } ) => {
+					// User patterns can have their sync statuses checked directly.
+					// Non-user patterns are all unsynced for the time being.
+					const [ , notSyncedStatus ] = SYNC_FILTERS;
+					return ! item.syncStatus
```
Comment (oandregal) [inline packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: What do you think of doing this instead:

```js
return ! item.syncStatus
    ? PATTERN_SYNC_TYPES.unsynced
    : SYNC_FILTERS.find( ( { value } ) => value === item.syncStatus )?.label;
```

From a reader point of view, I find it much clearer. It's also more robust because this won't break if anyone reorders the `SYNC_FILTERS` array, or adds a new item, etc.

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js [hunk 6]
```diff
@@ -235,12 +253,33 @@ export default function DataviewsPatterns() {
...
 					<Title item={ item } categoryId={ categoryId } />
 				),
-				maxWidth: 400,
 				enableHiding: false,
 			},
-		],
-		[ view.type, categoryId ]
-	);
+		];
+		if ( type === PATTERN_TYPES.theme ) {
```
Comment (oandregal) [inline packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: How do I trigger a situation in which the category type is `PATTERN_TYPES.user`?

Comment (oandregal) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: To test

- I went to the "manage my patterns" screen
- edited one of my patterns by adding a new "howdy" category

I see this new category is listed in the patterns page, though it works as any other:

https://github.com/WordPress/gutenberg/assets/583546/204b1c4b-5f0c-49cc-a8e5-718af1b0af93

Should I do anything differently?

Comment (oandregal) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: This is for later, but making `sync` visible and changing the listing hide the field (this is: the sidebar is not using the views yet, like the pages page).

https://github.com/WordPress/gutenberg/assets/583546/70496bfa-2388-4ebc-a26f-3d917e7468a6

Though, this makes me think that I wouldn't want the `sync` field to be removed from the view. Consider the following case: a user is navigating through the categories to see what patterns are available, though they are only interested in "not synced" patterns:

- `sync` field is shown in the layout
- `sync` filter is also shown

Captura de ecrã 2024-01-04, às 14 39 48

What happens when the user visits a user-provided category and then a theme provided category?

The consequences are that 1) visual jumps in the layout and 2) the user preferences (hiddenFields and filter statuses) are cleared once they visit a user-provided category.

It sounds it's to always render the `sync` field and its filter.

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js (hunk 5 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: `PATTERN_SYNC_TYPES` only contains the values and not the labels. Regarding the deconstruct I think it's simple and not worth it to have an extra loop there. Do you have strong opinions on using an extra `array.find`?

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js (hunk 6 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: >How do I trigger a situation in which the category type is PATTERN_TYPES.user?

Did you encountered something unexpected that is different from the old patterns page?
I'm not really sure but that param contains the value `wp_block`. Maybe @talldan or @aaronrobertshaw can shed some light here? 

>This is for later, but making sync visible and changing the listing hide the field (this is: the sidebar is not using the views yet, like the pages page).

We discussed this a bit with Riad [here](https://github.com/WordPress/gutenberg/pull/57488#discussion_r1440255470) and for now we reset the whole view. It does resemble the `custom` views we have in pages list. We can revisit later if we need to..

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js (hunk 5 above)
Comment (oandregal) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js]: My stronger concern is extracting the label from the deconstruct. This bit is fragile (others can add to that array, reorder it, etc.):

> const [ , notSyncedStatus ] = SYNC_FILTERS;

I'd be fine with almost any approach that doesn't use it.

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js (hunk 6 above)
Comment (oandregal) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: > Did you encountered something unexpected that is different from the old patterns page?

I don't know, because I don't know how to trigger a situation where type is not `PATTERN_TYPES.thems` :)

From what I gathered, my suggestion here is to always render the `sync-status` field and its filter.

Comment (ntsekouras) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: We definitely don't want to render it when `categoryType=wp_template_part`. I just used the same logic for displaying the filter as in the old page.

Comment (oandregal) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: > We definitely don't want to render it when categoryType=wp_template_part. I just used the same logic for displaying the filter as in the old page.

Oh, so this logic is not about theme vs. user, but about patterns vs. template parts? The naming was confusing. :)

It'd be best if we did `type !== TEMPLATE_PART_POST_TYPE`, though.

> I don't know, because I don't know how to trigger a situation where type is not PATTERN_TYPES.thems :)

I just realized this is triggered by selecting one of the template parts down below the sidebar. 👍

**Review by oandregal — APPROVED** [review]

File: packages/edit-site/src/components/page-patterns/dataviews-patterns.js (hunk 6 above)
Comment (oandregal) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: > We discussed this a bit with Riad https://github.com/WordPress/gutenberg/pull/57488#discussion_r1440255470 and for now we reset the whole view. It does resemble the custom views we have in pages list. We can revisit later if we need to..

I commented https://github.com/WordPress/gutenberg/pull/57488#discussion_r1441934166 For visibility, I think we should explore having categories as filter. And then re-implementing those new pages as "preset views" like we have for pages.

Comment (aaronrobertshaw) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: > Maybe @talldan or @aaronrobertshaw can shed some light here?

Originally, users couldn't create or manage their own pattern categories. Themes and core, however, added categories via `WP_Block_Pattern_Categories_Registry` and `register_block_pattern_category`.

To allow user-created patterns a taxonomy was used however it was a bit of an awkward fit in terms of maintaining backwards compatibility with the theme/core registered patterns. Eventually, those were able to be absorbed, so my current understanding is that on the patterns page they are essentially treated as the same thing now.

@glendaviesnz lead the efforts around pattern categorization and could provide more context or explanation if it's needed.

Comment (talldan) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: Taking a look. What's confusing is there are two `PATTERN_TYPES` consts that seem to have different values.

One is used in the inserter for inserter items:
https://github.com/WordPress/gutenberg/blob/6f61d8b802a880dc58eb490c17599cea0d614ba2/packages/block-editor/src/components/inserter/block-patterns-tab/utils.js#L7-L14

One is used by most of the other patterns code:
https://github.com/WordPress/gutenberg/blob/6f61d8b802a880dc58eb490c17599cea0d614ba2/packages/patterns/src/constants.js#L6-L9

Comment (talldan) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: <s>It seems the majority of the codebase uses the first variable, where `PATTERN_TYPES.user === 'user'`.

It seems you're using the second variable where `PATTERN_TYPES.user === 'wp_block'`, it doesn't seem to be used much in the codebase.

I think the first variable is the one you'll want to use, though I think it's still not great as the one object seems to mix and match lots of different concepts (syncing, pattern source, and I'm not sure what 'all' is).</s>

Ignore that, my assumption was wrong.

This definitely needs some improvement!

Comment (talldan) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: <s>So I think sometimes we're working with one kind of pattern item (where type is `user`) and other times a different type of item (where type is `wp_block`).</s>

I've started looking into improving the code quality. First PR here - [Rename patternBlock to patternPost](https://github.com/WordPress/gutenberg/pull/57568#top)

I've made another PR that renames the inserter constants to better indicate that they're for inserter items - [Code Quality: Improve inserter pattern constants](https://github.com/WordPress/gutenberg/pull/57570#top).

I think the goal should be to have a single normalized pattern item data structure (even if the data is sourced from different places) with consistent values, then hopefully we'll be able to tidy up the constants.

Comment (talldan) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: I think I have a full understanding of what the code is doing now.

There's a `categoryType` sprinkled across the codebase, but this is purely for routing and can be one of two values, `pattern` (for any time of pattern) or `wp_template_part` (for template parts). It's confusing that `PATTERN_TYPES` is often used to check this in addition to being used to check the pattern item types. The category type values are also unusual with one being a post type and the other not.

I'll follow up with another PR that improves this, but it'll be next week.

Comment (oandregal) [reply packages/edit-site/src/components/page-patterns/dataviews-patterns.js:259]: Nice, thanks for looking into this, Dan.
